git-worktree/src/lib.rs

```diff
@@ -1 +1,126 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_object::bstr::ByteSlice;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::time::Duration;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) {
+            from()
+            display("Could not convert path to UTF8: {}", err)
+        }
+        TimeError(err: std::time::SystemTimeError) {
+            from()
+            display("Could not read file time in proper format: {}", err)
+        }
+        ToU32Error(err: std::num::TryFromIntError) {
+            from()
+            display("Could not convert seconds to u32: {}", err)
+        }
+        IoError(err: std::io::Error) {
+            from()
+            display("IO Error: {}", err)
+        }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut git_index::State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<git_object::BlobRef<'a>>,
+{
+    let path = path.as_ref();
+    let mut buf = Vec::new();
+    let mut entry_time = Vec::new(); // Entries whose timestamps have to be updated
+    for (i, entry) in state.entries().iter().enumerate() {
+        if entry.flags.contains(git_index::entry::Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let dest = path.join(entry.path(state).to_path()?);
+        create_dir_all(dest.parent().expect("path is empty"))?;
```

Comment:
All over the codebase, one will write what's expected to be the case, instead of what went wrong. Here it would be more like `"entry paths are never empty"`.